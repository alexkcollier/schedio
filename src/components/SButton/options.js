export {
  colors,
  allShadeOptions,
  colorNames,
} from '@/utils/component-options'

export const sizes = [
  'small',
  'regular',
  'large',
]

export const types = [
  'outlined',
  'text',
]
