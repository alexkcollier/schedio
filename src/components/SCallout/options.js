export const calloutTypes = ['', 'success', 'warning', 'error']
