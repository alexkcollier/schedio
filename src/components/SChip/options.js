import { props } from '@/assets/styles/tokens/dist/tokens.raw.json'

export const colors = [
  ...Object.values(props).reduce((colorList, { category, name }) => {
    if (category === 'background-color') colorList.add(name.split('-')[0])

    return colorList
  }, new Set()),
].sort()
