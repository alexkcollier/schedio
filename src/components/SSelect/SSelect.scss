.select {
  @include control;
  @include control-with-icons;

  appearance: none;
  overflow: auto;

  &__icon--right {
    right: spacing('triple') !important;
  }

  &__container {
    &::after {
      @include arrow('down', 'right');

      pointer-events: none;
    }

    &--disabled {
      &::after {
        border-color: color('night', 'lighter');
      }
    }

    &--multiple {
      &::after {
        display: none;
      }

      .select__icon--right {
        right: spacing() + spacing('quarter') !important;
      }
    }
  }
}
