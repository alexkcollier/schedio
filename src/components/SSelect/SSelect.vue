<template>
  <SLabel
    :for="id"
    :label="label"
    :is-inline="isInline"
    :is-fullwidth="isFullwidth"
    :is-optional="!required"
    :show-optional="showOptional"
  >
    <div
      :class="{
        'select__container--multiple': multiple,
        'select__container--disabled': $attrs.disabled,
      }"
      class="select__container"
    >
      <select
        :id="id"
        v-model="selected"
        v-bind="$attrs"
        :class="{
          'select--invalid': isInvalid,
          'select--has-icon-left': !!iconLeft,
          'select--has-icon-right': !!iconRight
        }"
        :multiple="multiple"
        :name="name"
        :required="required"
        class="select"
        v-on="listeners"
      >
        <option
          :hidden="!multiple"
          :selected="!multiple"
          value=""
          disabled
        >
          {{ placeholder }}
        </option>

        <template v-if="optionsHaveGroups">
          <optgroup
            v-for="(group, groupName) in selectOptions"
            :key="groupName"
            :label="groupName"
          >
            <option
              v-for="(option, key) in group"
              :key="key"
              :value="getValue(option)"
            >
              {{ getLabel(option) }}
            </option>
          </optgroup>
        </template>

        <template v-else>
          <option
            v-for="(option, key) in selectOptions"
            :key="key"
            :value="getValue(option)"
          >
            {{ getLabel(option) }}
          </option>
        </template>
      </select>

      <SIcon
        v-if="iconLeft"
        :icon="iconLeft"
        :icon-color="iconLeftColor"
        class="select__icon select__icon--left"
      />

      <SIcon
        v-if="iconRight"
        :icon="iconRight"
        :icon-color="iconRightColor"
        class="select__icon select__icon--right"
      />
    </div>
  </SLabel>
</template>

<script>
import ControlIcons from '@/mixins/ControlIcons.mixin'
import InputText from '@/mixins/InputText.mixin'

export default {
  name: 'SSelect',

  mixins: [InputText, ControlIcons],

  props: {
    selectOptions: {
      type: [Object, Array],
      required: true
    },

    placeholder: {
      type: String,
      default: 'Select an option'
    },

    multiple: {
      type: Boolean,
      default: false
    }
  },

  data() {
    return {
      // <select multiple/> expects array
      selected: this.multiple ? [] : ''
    }
  },

  computed: {
    listeners() {
      return {
        ...this.$listeners,
        input: event => this.$emit('input', event.target.value),
        change: event => this.$emit('change', event.target.value)
      }
    },

    optionsHaveGroups() {
      if (!this.selectOptions) return false

      return Object.values(this.selectOptions).every(value => Array.isArray(value))
    }
  },

  methods: {
    getValue(option) {
      return option.value || option.label || option
    },
    getLabel(option) {
      return option.label || option.value || option
    }
  }
}
</script>
