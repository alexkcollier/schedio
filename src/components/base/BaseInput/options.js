export const allowed = ['email', 'text', 'password', 'url', 'search', ''].sort()
