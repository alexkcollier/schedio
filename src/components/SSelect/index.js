import SSelect from './SSelect.vue'

export { SSelect }

export default { install: Vue => Vue.component(SSelect.name, SSelect) }
