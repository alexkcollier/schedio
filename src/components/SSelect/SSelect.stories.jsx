import { SHeading } from '@/components/SHeading'
import { SSelect } from '@/components/SSelect'
import { withAttrsAsProps, withUnboundAttrs } from '@/mixins/stories/form-fields'
import PropList from '@/docs/PropList'
import StoryContainer from '@/docs/StoryContainer'
import { boolean, select, text, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'

storiesOf('Components/SSelect', module)
  .addDecorator(withKnobs)
  .add('SSelect', () => {
    const object1 = { label: 'Option 1', value: 'option1' }
    const object2 = { label: 'Option 2', value: 'option2' }
    const object3 = { label: 'Option 3', value: 'option3' }
    const object4 = { label: 'Option 4', value: 'option4' }
    const objectOfObjects = { object1, object2, object3, object4 }
    const arrayOfObjects = Object.values(objectOfObjects)
    const arrayOfStrings = arrayOfObjects.map(obj => obj.label)

    const labelledArraysOfObjects = {
      'Group 1': [{ label: 'Group 1 option label', value: 'Group 1 option value' }],
      'Group 2': [{ label: 'Group 2 option label', value: 'Group 2 option value' }],
      'Group 3': [{ label: 'Group 3 option label', value: 'Group 3 option value' }],
      'Group 4': [{ label: 'Group 4 option label', value: 'Group 4 option value' }]
    }

    const optionTypes = {
      'Object of objects': objectOfObjects,
      'Array of objects': arrayOfObjects,
      'Array of strings': arrayOfStrings,
      'Labelled arrays of objects': labelledArraysOfObjects
    }

    // enable setting of option structure.
    const optionKeys = Object.keys(optionTypes)
    const optionsObj = optionKeys.reduce((o, k) => ({ ...o, [k]: k }), {})
    const options = select('selectOptions', optionsObj, optionKeys[0], 'Required Props')

    return {
      props: {
        props: {
          default: {
            ...withAttrsAsProps({
              id: 'base-select',
              name: 'base-select',
              label: 'Select'
            }),
            placeholder: text('Placeholder', 'Select an option', 'Optional Props'),
            selectOptions: optionTypes[options],
            multiple: boolean('Multiple', false, 'Optional Props'),
            isInline: boolean('isInline', false, 'Optional Props'),
            isFullwidth: boolean('isFullwidth', false, 'Optional Props')
          }
        },
        attrs: {
          default: withUnboundAttrs()
        }
      },
      render(h) {
        const { props, attrs } = this.$props

        return (
          <StoryContainer>
            <SHeading level="1">Select</SHeading>
            <p>
              Selects provide a list of options to choose from. They can be set to allow selection
              of none, one or many options.
            </p>
            <p>A select's options can have the following structures:</p>
            <ul>
              {optionKeys.map(key => (
                <li>{key}</li>
              ))}
            </ul>

            <SHeading level="2">Example</SHeading>
            <SSelect {...{ props }} {...{ attrs }} />

            <SHeading level="3" style="text-transform: none;">
              selectOptions:
            </SHeading>
            <pre>{JSON.stringify(props.selectOptions, '', 2)}</pre>

            <PropList component={SSelect} />
          </StoryContainer>
        )
      }
    }
  })
