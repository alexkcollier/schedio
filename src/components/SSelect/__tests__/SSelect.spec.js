import { colors as iconColors } from '@/components/SIcon/options'
import SSelect from '@/components/SSelect/SSelect.vue'
import { shallowMount } from '@vue/test-utils'
import icons from 'feather-icons/dist/icons.json'

const iconList = Object.keys(icons)

describe('SSelect.vue', () => {
  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => {})
  const inputListener = jest.fn()
  const defaultOptions = ['one', 'two']

  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(SSelect, {
      propsData: {
        id: 'select',
        name: 'select',
        label: 'Select',
        selectOptions: defaultOptions
      },
      listeners: {
        change: inputListener,
        input: inputListener
      }
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('requires required props', () => {
    shallowMount(SSelect)
    expect(errorSpy).toBeCalled()
  })

  it('can have a single placeholder', () => {
    const placeholder = 'This is a placeholder'

    wrapper.setProps({ placeholder })
    expect(wrapper.text()).toContain(placeholder)

    const placeholderOptions = wrapper.findAll('option').filter(w => w.text() === placeholder)

    expect(placeholderOptions.length).toBe(1)
    expect(placeholderOptions.at(0).value).toBeUndefined()
    expect(placeholderOptions.at(0).element.disabled).toBeTruthy()
    expect(placeholderOptions.at(0).text()).toBe(placeholder)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can use an array of objects for `selectOptions`', () => {
    wrapper.setProps({ selectOptions: [{ label: 'One', value: 1 }, { label: 'Two', value: 2 }] })
    expect(wrapper.findAll('option').length).toBeGreaterThan(1)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can use an object of objects for `selectOptions`', () => {
    wrapper.setProps({
      selectOptions: {
        optionOne: { label: 'One', value: 1 },
        optionTwo: { label: 'Two', value: 2 }
      }
    })
    expect(wrapper.findAll('option').length).toBeGreaterThan(1)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can use an object of arrays `selectOptions` with groups', () => {
    const optGroups = {
      groupOne: [{ label: 'One', value: 1 }],
      groupTwo: [{ label: 'Two', value: 2 }]
    }

    wrapper.setProps({ selectOptions: optGroups })
    expect(wrapper.findAll('option').length).toBeGreaterThan(1)
    expect(wrapper.contains('optgroup')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be a multiple select', () => {
    wrapper.setProps({ multiple: true })
    expect(wrapper.find('select').element.multiple).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('accepts input', () => {
    wrapper.find('select').setValue(defaultOptions[0])
    expect(inputListener).toBeCalled()
    expect(wrapper.emitted('change')[0][0]).toBe(defaultOptions[0])
  })

  it('can be invalid', () => {
    wrapper.setProps({ isInvalid: true })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be required', () => {
    wrapper.setProps({ required: true })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be inline', () => {
    wrapper.setProps({ isInline: true })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be fullwidth', () => {
    wrapper.setProps({ isFullwidth: true })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can have an icon on the left', () => {
    wrapper.setProps({ iconLeft: iconList[0] })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can have an icon on the right', () => {
    wrapper.setProps({ iconRight: iconList[0] })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can set each icon color', () => {
    wrapper.setProps({
      iconLeftColor: iconColors[0],
      iconRightColor: iconColors[1]
    })

    expect(errorSpy).not.toBeCalled()
    expect(wrapper.props('iconLeftColor')).not.toBe(wrapper.props('iconRightColor'))
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can hide the optional tag', () => {
    wrapper.setProps({ showOptional: false, required: true })
    expect(wrapper.text()).not.toContain('Optional')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be disabled correctly', () => {
    wrapper.setProps({ disabled: 'disabled' })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
