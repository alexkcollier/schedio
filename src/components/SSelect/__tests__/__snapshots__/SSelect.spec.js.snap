// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSelect.vue can be a multiple select 1`] = `
<slabel-stub label="Select" showoptional="true" for="select">
  <div class="select__container select__container--multiple"><select id="select" name="select" required="required" class="select" multiple="multiple">
      <option value="" disabled="disabled">
        Select an option
      </option>
      <option value="one">
        one
      </option>
      <option value="two">
        two
      </option>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can be disabled correctly 1`] = `
<slabel-stub label="Select" showoptional="true" for="select">
  <div class="select__container select__container--disabled"><select id="select" name="select" required="required" class="select" disabled="disabled">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <option value="one">
        one
      </option>
      <option value="two">
        two
      </option>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can be fullwidth 1`] = `
<slabel-stub label="Select" showoptional="true" for="select" isfullwidth="true">
  <div class="select__container"><select id="select" name="select" required="required" class="select">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <option value="one">
        one
      </option>
      <option value="two">
        two
      </option>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can be inline 1`] = `
<slabel-stub label="Select" showoptional="true" for="select" isinline="true">
  <div class="select__container"><select id="select" name="select" required="required" class="select">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <option value="one">
        one
      </option>
      <option value="two">
        two
      </option>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can be invalid 1`] = `
<slabel-stub label="Select" showoptional="true" for="select">
  <div class="select__container"><select id="select" name="select" required="required" class="select select--invalid">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <option value="one">
        one
      </option>
      <option value="two">
        two
      </option>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can be required 1`] = `
<slabel-stub label="Select" showoptional="true" for="select">
  <div class="select__container"><select id="select" name="select" required="required" class="select">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <option value="one">
        one
      </option>
      <option value="two">
        two
      </option>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can have a single placeholder 1`] = `
<slabel-stub label="Select" showoptional="true" for="select">
  <div class="select__container"><select id="select" name="select" required="required" class="select">
      <option hidden="hidden" value="" disabled="disabled">
        This is a placeholder
      </option>
      <option value="one">
        one
      </option>
      <option value="two">
        two
      </option>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can have an icon on the left 1`] = `
<slabel-stub label="Select" showoptional="true" for="select">
  <div class="select__container"><select id="select" name="select" required="required" class="select select--has-icon-left">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <option value="one">
        one
      </option>
      <option value="two">
        two
      </option>
    </select>
    <sicon-stub icon="activity" iconcolor="green" iconsize="" class="select__icon select__icon--left"></sicon-stub>
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can have an icon on the right 1`] = `
<slabel-stub label="Select" showoptional="true" for="select">
  <div class="select__container"><select id="select" name="select" required="required" class="select select--has-icon-right">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <option value="one">
        one
      </option>
      <option value="two">
        two
      </option>
    </select>
    <!---->
    <sicon-stub icon="activity" iconcolor="green" iconsize="" class="select__icon select__icon--right"></sicon-stub>
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can hide the optional tag 1`] = `
<slabel-stub label="Select" for="select">
  <div class="select__container"><select id="select" name="select" required="required" class="select">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <option value="one">
        one
      </option>
      <option value="two">
        two
      </option>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can set each icon color 1`] = `
<slabel-stub label="Select" showoptional="true" for="select">
  <div class="select__container"><select id="select" name="select" required="required" class="select">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <option value="one">
        one
      </option>
      <option value="two">
        two
      </option>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can use an array of objects for \`selectOptions\` 1`] = `
<slabel-stub label="Select" showoptional="true" for="select">
  <div class="select__container"><select id="select" name="select" required="required" class="select">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <option value="1">
        One
      </option>
      <option value="2">
        Two
      </option>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can use an object of arrays \`selectOptions\` with groups 1`] = `
<slabel-stub label="Select" showoptional="true" for="select">
  <div class="select__container"><select id="select" name="select" required="required" class="select">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <optgroup label="groupOne">
        <option value="1">
          One
        </option>
      </optgroup>
      <optgroup label="groupTwo">
        <option value="2">
          Two
        </option>
      </optgroup>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue can use an object of objects for \`selectOptions\` 1`] = `
<slabel-stub label="Select" showoptional="true" for="select">
  <div class="select__container"><select id="select" name="select" required="required" class="select">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <option value="1">
        One
      </option>
      <option value="2">
        Two
      </option>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;

exports[`SSelect.vue renders correctly 1`] = `
<slabel-stub label="Select" showoptional="true" for="select">
  <div class="select__container"><select id="select" name="select" required="required" class="select">
      <option hidden="hidden" value="" disabled="disabled">
        Select an option
      </option>
      <option value="one">
        one
      </option>
      <option value="two">
        two
      </option>
    </select>
    <!---->
    <!---->
  </div>
</slabel-stub>
`;
